.claude/skills/add-cuda-kernel/skill.md

```diff
@@ -0,0 +1,575 @@
+- Export using `TVM_FFI_DLL_EXPORT_TYPED_FUNC(name, function)`
+
+**TVM-FFI Error Handling:**
+
+- `TVM_FFI_THROW(ValueError) << "message"` - Throw ValueError with custom message
+- `TVM_FFI_THROW(TypeError) << "message"` - Throw TypeError
+- Use `<<` to chain multiple values in the error message
+- Errors are properly propagated back to Python
+
+## Step 4: Create JIT Generator (No Jinja for Simple Case)
+
+Create `flashinfer/jit/scale.py`:
+
+```python
+import os
+import shutil
+from pathlib import Path
+
+from . import JitSpec, gen_jit_spec
+from . import env as jit_env
+from .core import write_if_different
+
+
+def get_scale_uri(dtype_in: str, dtype_out: str) -> str:
+    """Generate unique identifier for scale module."""
+    return f"scale_dtype_in_{dtype_in}_dtype_out_{dtype_out}"
+
+
+def gen_scale_module(dtype_in, dtype_out):
+    """
+    Generate JIT module for scale operation.
+
+    Note: This is a simple example without Jinja templating.
+    The dtype dispatch is handled at runtime in the C++ code.
+    """
+    # Compute URI
+    uri = get_scale_uri(dtype_in, dtype_out)
+
+    # Create generation directory
+    gen_directory = jit_env.FLASHINFER_GEN_SRC_DIR / uri
+    os.makedirs(gen_directory, exist_ok=True)
+
+    # Copy source files (no Jinja needed for this simple case)
+    sources = []
+    for fname in ["scale.cu", "scale_jit_binding.cu"]:
+        src_path = jit_env.FLASHINFER_CSRC_DIR / fname
+        dest_path = gen_directory / fname
+        shutil.copy(src_path, dest_path)
+        sources.append(dest_path)
+
+    # Return JitSpec
+    return gen_jit_spec(
+        name=uri,
+        sources=sources,
+        extra_cuda_cflags=[],
+    )
+```
+
+**Key points:**
+
+- No Jinja template needed for simple operations
+- Just copy source files to generation directory
+- URI uniquely identifies the module configuration
+
+### (Optional) Specifying Supported CUDA Architectures
+
+FlashInfer uses `CompilationContext` to manage CUDA architecture targets. This is critical because some kernels only work on specific GPU architectures (e.g., Hopper SM90, Blackwell SM100).
+
+#### How CompilationContext Works
+
+**Automatic Detection** (default):
+```python
+from flashinfer.compilation_context import CompilationContext
+
+ctx = CompilationContext()
+# Automatically detects all GPUs in the system
+# For SM90+, adds 'a' suffix (e.g., 9.0a for Hopper)
+# Result: ctx.TARGET_CUDA_ARCHS = {(9, '0a'), (10, '0a'), ...}
+```
+
+**Manual Override** (via environment variable):
+```bash
+export FLASHINFER_CUDA_ARCH_LIST="8.0 9.0a 10.0a"
+# Now only these architectures will be compiled
+```
+
+#### Specifying Architectures in Your JIT Module
+
+When creating a JIT module, specify which major SM versions are supported:
+
+```python
+from flashinfer.jit.core import gen_jit_spec
+from flashinfer.jit import current_compilation_context
+
+def gen_my_hopper_only_module():
+    """Example: Kernel only works on SM90+ (Hopper and newer)"""
```

Comment:
[nit] We want to be careful with statements like "SM90+". We are not guaranteeing backwards compatibility for e.g., SM203.